http/src/lib.rs

```diff
@@ -235,78 +182,177 @@ impl<M: jsonrpc::Metadata, S: jsonrpc::Middleware<M>> ServerBuilder<M, S> {
+			new_hosts.insert(address.clone());
+			new_hosts.insert(address.replace("127.0.0.1", "localhost"));
+			new_hosts.into_iter().collect()
+		})
+	}
+
 	/// Start this JSON-RPC HTTP server trying to bind to specified `SocketAddr`.
 	pub fn start_http(self, addr: &SocketAddr) -> ServerResult {
-		let panic_for_server = Arc::new(Mutex::new(self.panic_handler));
 		let cors_domains = self.cors_domains;
-		let hosts = Arc::new(Mutex::new(self.allowed_hosts));
-		let hosts_setter = hosts.clone();
-
-		let (handler, event_loop) = match self.jsonrpc_handler {
-			Handler::WithLoop(handler) => (handler, None),
-			Handler::Simple(io_handler) => {
-				let event_loop = RpcEventLoop::spawn();
-				(event_loop.handler(Arc::new(io_handler)), Some(event_loop.into()))
-			},
-		};
-
-		let jsonrpc_handler = Rpc {
-			handler: handler,
-			extractor: self.meta_extractor,
-		};
-
-		let (l, srv) = try!(try!(hyper::Server::http(addr)).handle(move |control| {
-			let handler = PanicHandler { handler: panic_for_server.clone() };
-			let hosts = hosts.lock().unwrap().clone();
-			ServerHandler::new(jsonrpc_handler.clone(), cors_domains.clone(), hosts, handler, control)
-		}));
-
-		// Add current host to allowed headers.
-		// NOTE: we need to use `l.addrs()` instead of `addr`
-		// it might be different!
-		{
-			let mut hosts = hosts_setter.lock().unwrap();
-			if let Some(current_hosts) = hosts.take() {
-				let mut new_hosts = current_hosts.into_iter().collect::<HashSet<_>>();
-				for addr in l.addrs() {
-					let address = addr.to_string();
-					new_hosts.insert(address.clone());
-					new_hosts.insert(address.replace("127.0.0.1", "localhost"));
-				}
-				// Override hosts
-				*hosts = Some(new_hosts.into_iter().collect());
-			}
-		}
-
+		let allowed_hosts = self.allowed_hosts;
+		let handler = self.jsonrpc_handler;
+		let meta_extractor = self.meta_extractor;
+		let panic_handler = self.panic_handler;
+		let threads = self.threads;
+
+		let (local_addr_tx, local_addr_rx) = mpsc::channel();
+		let (close, shutdown_signal) = futures::sync::oneshot::channel();
+		let addr = addr.to_owned();
 		let handle = thread::spawn(move || {
-			srv.run();
+			let _panic_handler = PanicHandler { handler: panic_handler };
+
+			// TODO [ToDr] Errors?
+			let hosts = Arc::new(RwLock::new(allowed_hosts.clone()));
+			let hosts2 = hosts.clone();
+
+			let mut server = tokio_proto::TcpServer::new(tokio_minihttp::Http, addr);
+			server.threads(threads);
+			let server = server.bind(move || Ok(RpcService {
+				handler: handler.clone(),
+				meta_extractor: meta_extractor.clone(),
+				hosts: hosts.read().clone(),
+				cors_domains: cors_domains.clone(),
+			})).expect("Cannot bind to socket.");
+
+			let local_addr = server.local_addr().expect("OK");
+			// Add current host to allowed headers.
+			// NOTE: we need to use `local_address` instead of `addr`
+			// it might be different!
+			*hosts2.write() = Self::update_hosts(allowed_hosts, local_addr.clone());
```

Comment:
The problem is that we can only learn what address is we're listening on after we call `bind` which requires the closure.

There is no risk of race condition here, cause the server is just bound to an interface, but is not started yet, so the closure is never invoked before we update the `hosts`.